server/src/main/java/org/elasticsearch/transport/InboundHandler.java

```diff
@@ -106,70 +106,98 @@ private void messageReceived(TcpChannel channel, InboundMessage message, long st
-                long requestId = header.getRequestId();
-                if (header.isHandshake()) {
-                    responseHandler = handshaker.removeHandlerForHandshake(requestId);
-                } else {
-                    final TransportResponseHandler<? extends TransportResponse> theHandler = responseHandlers.onResponseReceived(
-                        requestId,
-                        messageListener
-                    );
-                    if (theHandler == null && header.isError()) {
-                        responseHandler = handshaker.removeHandlerForHandshake(requestId);
-                    } else {
-                        responseHandler = theHandler;
-                    }
-                }
+                responseHandler = findResponseHandler(header);
                 // ignore if its null, the service logs it
                 if (responseHandler != null) {
-                    if (message.getContentLength() > 0 || header.getVersion().equals(TransportVersion.current()) == false) {
-                        final StreamInput streamInput = namedWriteableStream(message.openOrGetStreamInput());
-                        assertRemoteVersion(streamInput, header.getVersion());
-                        if (header.isError()) {
-                            handlerResponseError(streamInput, message, responseHandler);
-                        } else {
-                            handleResponse(remoteAddress, streamInput, responseHandler, message);
-                        }
-                    } else {
-                        assert header.isError() == false;
-                        handleResponse(remoteAddress, EMPTY_STREAM_INPUT, responseHandler, message);
-                    }
+                    resolveResponseHandler(message, responseHandler, remoteAddress);
                 }
             }
         } finally {
             final long took = threadPool.rawRelativeTimeInMillis() - startTime;
             handlingTimeTracker.addHandlingTime(took);
             final long logThreshold = slowLogThresholdMs;
             if (logThreshold > 0 && took > logThreshold) {
-                if (isRequest) {
-                    logger.warn(
-                        "handling request [{}] took [{}ms] which is above the warn threshold of [{}ms]",
-                        message,
-                        took,
-                        logThreshold
-                    );
-                } else {
-                    logger.warn(
-                        "handling response [{}] on handler [{}] took [{}ms] which is above the warn threshold of [{}ms]",
-                        message,
-                        responseHandler,
-                        took,
-                        logThreshold
-                    );
-                }
+                logSlowMessage(message, took, logThreshold, responseHandler);
+            }
+        }
+    }
+
+    private void resolveResponseHandler(
```

Comment:
nit: `resolve` suggests to me we're doing some kind of lookup to find the response handler, but in fact we already have the handler and this method is running it. Which admittedly could be called a "resolution". But still, I'd rather name this something like `completeResponseHandler` or `executeResponseHandler`.